### ISSUE TITLE
Maximizing inference speed with good quality predictions for multi-modal and multi-label task

### ISSUE THREAD

**Issue by ijmiller2** [awslabs/autogluon#1540]
Hey there,

First of all thank you for this amazing resource. I have found the documentation to be very helpful and the code to be quite stable across difference machines.

I'm working on a multi-modal (image + raw text + structured) dataset with a multi-label task. I'm trying to understand the logic autogluon uses pick model types given the time limit and presets. Ideally I'd like to incorporate the image predictor (in my experience this is the most predictive and generalizable feature for this dataset), but depending on which parameters I use it seems to often exclude that type of Predictor. If I raise the time limit substantially it seems to include them, but then also seems to add significantly more stacking and bagging which increases train time (which is okay) but I worry that it will slow down inference speed substantially. While I definitely have uses for a more complex ensemble, I am currently trying to optimize for real time predictions (so hoping for sub-second latency for predictions). I know there's an argument for "exclude_model_types", is there or could there be an equivalent to "include_predictor_types" for the multi-modal application?

So far I have tried the recommended "good_quality_faster_inference_only_refit" and "optimize_for_deployment" presets. But they seem to exclude the ImagePredictor unless I substantially increase the time limit. Which by the way, doesn't seem to be respected (or at least handled consistently) in the multi-label setting (perhaps it's the time limit for each feature / column in the multi-label case?). But then when I increase the time limit, the amount of bagging and stacking appears to increase as well. Again my main concern with that behavior is inference times that are not practical for real-time applications (for a mobile app that powers user experience).

Thanks in advance for any advice / insight.

Best,
Ian

**Comment by Innixma** [comment]
Hi Ian,

Thanks for opening this issue! It seems you are trying to solve a very challenging problem that incorporates a lot of advanced functionality in AutoGluon, such as multi-modality and multi-label, while also having strict constraints on the inference speed of the final result.

Regarding time limit, time limit will be for each label in multi-label, for example if you specify 300 seconds, and you have 5 labels, training will take ~1500 seconds.

Regarding inference time concerns with bagging, `good_quality_faster_inference_only_refit` will not increase inference time due to bagging as it collapses all bags at end of training into single models. If you provide enough training time for all models to train to completion and it still excludes the image model, that means that the image model is not strong enough to benefit the ensemble, and thus is not necessary.

Our upcoming v0.4 release (~End of Feb) may help significantly in both performance and inference. We have an in-progress PR #1537 which refactors TextPredictor and also adds a new experimental AutoMMPredictor that handles ITT (image, text, tabular) data. This new experimental predictor might be able to satisfy your inference constraints while producing strong results. It should be available via TabularPredictor interface via the `multimodal` hyperparameter preset in v0.4 (assuming we have time to test it and it is working well, this is a very new area of research).

**Comment by ijmiller2** [comment]
Hi Nick,

Thank you for the thoughtful and detailed response! I suppose I am working on a more complicated than average task and the low-latency aspirations are somewhere between ambitious to unreasonable haha. Though I realize I should maybe clarify that the hope for sub-second latency is for a single prediction (and not a big batch of them), which is what I currently get from simpler (non-ensemble and mostly image classification) models.

Good to know about the time limit for multi-label models and that the `good_quality_faster_inference_only_refit` will not increase inference time. That's very helpful - thanks for clarifying!

I think what might be happening with the image model being left out has to do with the models overfitting the text-based features. Though these text based features in the training (and validation / test data) are very descriptive, models that rely too much on them tend not to generalize well to the type of text features seen out in the wild (where the model is being used to classify new things that have much more variance in this feature type). But this is mostly my own problem that probably requires more preprocessing and engineering on my side. So I'm also working on dealing with that independently.

Very exciting to hear about the upcoming release and updates to multimodal functionalities! Let me know if there's any way I can help to test these. I'm guessing you guys have your own suite of tools to monitor system performance (and I know there are many others out there), but I have personally found [Neptune.ai's](https://github.com/neptune-ai) suite of tools to be quite useful. For instance, attaching a screenshot from a recent Autogluon run with multi-modal and multi-label data. What is interesting here is that the memory profile increases in a step-wise manner even though I have `consider_labels_correlation` set to `False`. I started with it set to `True` (since I do expect some inter-dependence of labels here), but then after seeing the initially memory profile and considering that such a task would indeed require much more memory, I set it to `False`, expecting it to be more stable/flat.

![image](https://user-images.githubusercontent.com/8560151/154533979-7ec1f955-2d19-47ed-8967-9d1c79f4df4f.png)

On a somewhat related note, after training one model (on a subset of data (and labels); with ~1500 samples and 9 labels) and presets including `'keep_only_best': True, 'save_space': True`, the output directory was about ~12.5 GB in size, with most of that size appearing to come from TextPredictor(s) (about 3 GB each) from 4/9 of those labels.

Thanks again for all your work and help here,
Ian
